File: airbyte-commons-temporal/src/main/java/io/airbyte/commons/temporal/scheduling/retries/BackoffPolicy.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.temporal.scheduling.retries;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import java.time.Duration;
import lombok.Builder;
import lombok.Getter;
import lombok.ToString;
import lombok.extern.jackson.Jacksonized;

/**
 * Calculates how much we should wait based off the provided configuration. Uses exponential backoff
 * clamped to a min/max interval.
 */
@Builder
@Getter
@ToString
@Jacksonized
@JsonIgnoreProperties(ignoreUnknown = true)
public class BackoffPolicy {

  @Builder.Default
  private Duration minInterval = Duration.ofSeconds(10);
  @Builder.Default
  private Duration maxInterval = Duration.ofHours(1);
  @Builder.Default
  private long base = 2;

  /**
   * Calculates backoff based off provided steps.
   *
   * @param ordinal the number of adverse effects we have observed.
   * @return backoff duration to wait.
   */
  @SuppressWarnings("PMD.AvoidLiteralsInIfCondition")
  public Duration getBackoff(final long ordinal) {
    if (ordinal < 1) {
      return Duration.ZERO;
    }

    final var coefficient = Math.pow(base, ordinal - 1);
    final var calculated = (long) (minInterval.toMillis() * coefficient);
    final var clamped = Math.min(calculated, maxInterval.toMillis());

    return Duration.ofMillis(clamped);
  }

}

File: airbyte-commons-temporal/src/main/kotlin/io/airbyte/commons/temporal/scheduling/retries/BackoffPolicy.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.commons.temporal.scheduling.retries

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import java.time.Duration
import kotlin.math.min
import kotlin.math.pow

private val DEFAULT_BASE: Long = 2

/**
 * Calculates how much we should wait based off the provided configuration. Uses exponential backoff
 * clamped to a min/max interval.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
data class BackoffPolicy(
  val minInterval: Duration = Duration.ofSeconds(10),
  val maxInterval: Duration = Duration.ofHours(1),
  val base: Long = DEFAULT_BASE,
) {
  constructor(minInterval: Duration, maxInterval: Duration) : this(minInterval = minInterval, maxInterval = maxInterval, base = DEFAULT_BASE)

  /**
   * Calculates backoff based off provided steps.
   *
   * @param ordinal the number of adverse effects we have observed.
   * @return backoff duration to wait.
   */
  fun getBackoff(ordinal: Long): Duration {
    if (ordinal < 1) {
      return Duration.ZERO
    }

    val coefficient = base.toDouble().pow((ordinal - 1).toDouble())
    val calculated = (minInterval.toMillis() * coefficient).toLong()
    val clamped = min(calculated.toDouble(), maxInterval.toMillis().toDouble()).toLong()

    return Duration.ofMillis(clamped)
  }
}
